application/apps/text_grep/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+use grep_regex::RegexMatcher;
+use grep_searcher::sinks::UTF8;
+use grep_searcher::Searcher;
+use rayon::prelude::*;
+use std::collections::HashMap;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+use std::sync::mpsc::Sender;
+use std::sync::{mpsc, Arc, Mutex};
+use std::thread;
+use std::time::Instant;
+use tokio_util::sync::CancellationToken;
+
+#[derive(Debug)]
+pub struct SearchResult {
+    pub file_path: String,
+    pub pattern_counts: HashMap<String, usize>,
+    pub error_message: Option<String>,
+}
+
+pub struct TextGrep;
+
+impl Default for TextGrep {
+    fn default() -> Self {
+        TextGrep::new()
+    }
+}
+
+impl TextGrep {
+    pub fn new() -> Self {
+        TextGrep
+    }
+
+    pub async fn count_occurrences(
+        &self,
+        patterns: Vec<&str>,
+        file_paths: Vec<&str>,
+        chunk_size: usize,
+        cancel_token: CancellationToken,
+    ) -> Result<Vec<SearchResult>, String> {
+        let mut results = Vec::new();
+        let cancel_token_clone = cancel_token.clone();
+
+        let (sender, receiver) = mpsc::channel();
+        let (error_sender, error_receiver) = mpsc::channel();
+
+        let patterns_arc: Vec<_> = patterns.iter().map(|&p| Arc::from(p)).collect();
+        let file_paths_arc: Vec<_> = file_paths.iter().map(|&fp| Arc::from(fp)).collect();
+
+        let thread_handles: Vec<_> = file_paths_arc
+            .par_iter()
+            .map(|file_path| {
+                let patterns = patterns_arc.clone();
+                let sender = sender.clone();
+                let error_sender_clone = error_sender.clone();
+                let cancel_token = cancel_token_clone.clone();
+                let file_path = Arc::clone(file_path);
+                thread::spawn(move || {
```

Comment:
why we need to spawn system thread with async method? In general it's very expensive. If we really need threads here tokio's `task` should be used. But I don't think we need it at all.